src/ws.nim

```diff
@@ -0,0 +1,570 @@
+  MaxHttpRequestSize = 128 * 1024  # maximum size of HTTP body in octets
+  HttpHeadersTimeout = timer.seconds(120) # timeout for receiving headers (120 sec)
+  HttpBodyTimeout = timer.seconds(12)     # timeout for receiving body (12 sec)
+  HeadersMark = @[byte(0x0D), byte(0x0A), byte(0x0D), byte(0x0A)]
+  WebsocketUserAgent* = "nim-ws (https://github.com/status-im/nim-ws)"
+
+type
+  ReadyState* = enum
+    Connecting = 0 # The connection is not yet open.
+    Open = 1       # The connection is open and ready to communicate.
+    Closing = 2    # The connection is in the process of closing.
+    Closed = 3     # The connection is closed or couldn't be opened.
+
+  WebSocket* = ref object
+    tcpSocket*: StreamTransport
+    version*: int
+    key*: string
+    protocol*: string
+    readyState*: ReadyState
+    masked*: bool # send masked packets
+    header*: HttpHeaders
+
+  AsyncCallback = proc (transp: StreamTransport,
+      header: HttpRequestHeader): Future[void] {.closure, gcsafe.}
+  HttpServer* = ref object of StreamServer
+    callback: AsyncCallback
+
+  ReqStatus = enum
+    Success, Error, ErrorFailure
+
+  WebSocketError* = object of IOError
+
+  HttpHeaders* = ref object
+    table*: TableRef[string, seq[string]]
+
+type
+  AsyncHttpClient* = ref object
+    connected: bool
+    currentURL: Url          ## Where we are currently connected.
+    headers: seq[HttpHeader] ## Headers to send in requests.
+    userAgent: string
+    contentTotal: BiggestInt
+    contentProgress: BiggestInt
+    oneSecondProgress: BiggestInt
+    transp: StreamTransport
+    buf: seq[byte]
+
+template `[]`(value: uint8, index: int): bool =
+  ## Get bits from uint8, uint8[2] gets 2nd bit.
+  (value and (1 shl (7 - index))) != 0
+
+proc nibbleFromChar(c: char): int =
+  ## Converts hex chars like `0` to 0 and `F` to 15.
+  case c:
+    of '0'..'9': (ord(c) - ord('0'))
+    of 'a'..'f': (ord(c) - ord('a') + 10)
+    of 'A'..'F': (ord(c) - ord('A') + 10)
+    else: 255
+
+proc nibbleToChar(value: int): char =
+  ## Converts number like 0 to `0` and 15 to `fg`.
+  case value:
+    of 0..9: char(value + ord('0'))
+    else: char(value + ord('a') - 10)
+
+proc decodeBase16*(str: string): string =
+  ## Base16 decode a string.
+  result = newString(str.len div 2)
+  for i in 0 ..< result.len:
+    result[i] = chr(
+      (nibbleFromChar(str[2 * i]) shl 4) or
+      nibbleFromChar(str[2 * i + 1]))
+
+proc encodeBase16*(str: string): string =
+  ## Base61 encode a string.
+  result = newString(str.len * 2)
+  for i, c in str:
+    result[i * 2] = nibbleToChar(ord(c) shr 4)
+    result[i * 2 + 1] = nibbleToChar(ord(c) and 0x0f)
+
+proc genMaskKey(): array[4, char] =
+  ## Generates a random key of 4 random chars.
+  proc r(): char = char(rand(255))
+  [r(), r(), r(), r()]
+
+proc handshake*(ws: WebSocket, header: HttpRequestHeader) {.async.} =
+  ## Handles the websocket handshake.
+  ws.version = parseInt(header["Sec-WebSocket-Version"])
```

Comment:
std.parseInt() should not be used, because it raises too many exceptions and allows negative values like `-1`.